cpu/atmega_common/periph/pm.c

```diff
@@ -28,11 +28,18 @@
 
 void pm_reboot(void)
 {
+    #if defined(CPU_ATMEGA256RFR2)
+    /* set a variable to stack to signalizes soft reset */
+    /* clear MCU Status Register Interrupt flags */
+    MCUSR = 0x00;
+    __asm__ __volatile__("mov r3, %0\n" :: "r" (0xAA));
+    #endif /* CPU_ATMEGA256RFR2 */
```

Comment:
Is this platform the only avr8 that uses this trick to realize the "soft reset"? On first sight is seems unattractive to implement it here. Any chance to implement this on the CPU basis and call a function in worst case?

Reply:
All other platforms can not distinguish WDT reset and software triggered WDT reset.
They use software reset but later on they will not be able to recognize a real WDT reset.

I don't like the idea to wrap an additional function around pm_reboot. As it then also needs changes in the shell etc.

They could adapt it.

Reply:
AFAIK AVR doesn't have a properly named "software reset" so we did the trick with the WDT (on all AVR8 platforms, afaik). Since we don't have a WDT API yet, I suggest leave this as is and modify it accordingly when the WDT API is ready. No platform in my knowledge uses any WDT in RIOT, especially ARM on which once enabled it's impossible to disable it.

Reply:
I don't understand why an WDT api is neccessary to distinguish soft and hardware reset? This is just the proper way IMHO to do a WDT based software reset, without affecting a WDT api regardless when it gets implemented.

Reply:
This is not a standard thing. We saved the MCUSR register within the bootloader as it will write MCUSR to 0. Since it is saved we are able to retrieve the information in the application and check if the reset was issued by the application. Without setting MCUSR to 0, the application would always report a Watchdog reset, as the bootloader uses the watchdog. Maybe the board would be a better place for a function like this, as it is bootloader specific and not CPU specific.

Reply:
Please don't mix this things.
1. with some devices it is necessary to reset the WDT as fast as possible. "the watchdog timer remains active even after a system reset (except a power-on condition), using the fastest prescaler value (approximately 15 ms). It is therefore required to turn off the watchdog early during program startup, the datasheet recommends a sequence like the following:"

https://www.microchip.com/webdoc/AVRLibcReferenceManual/group__avr__watchdog.html

While implementing this i saw if the bootloader resets MCUSR the application will not be able to get reset information. Thus we save MCUSR to "r2".

and retrieve it in .init0.

2. While doing so it was not far fetch to also implement a feature to recognize software reset.
Which is not connected to the WDT reset error which will occur if MCSUR is not cleared within 15ms.

Reply:
> I don't understand why an WDT api is neccessary to distinguish soft and hardware reset?

What I mean is that in this case "it doesn't matter" if you use the watchdog for a software reset, since the watchdog functionality is anyways not used at all in RIOT.

Reply:
Yet, and the one who implements the WDT api should then know out of nowhere that we are missusing the WDT and thererfore than implement a functionality to avoid wrong detection?

As we use the WDT as something defferent i think we should also take care that whom ever wants to implement the WDT will not stumble over half implemented features. IMHO.